### ISSUE TITLE
Inaccurate white balance

### ISSUE THREAD

**Issue by Vandermode** [aasharma90/Unprocessing_for_RAW_Denoising#2]
Hi, thank you for your nice reimplementation.

I was wondering whether you have figured out the inaccurate white balance issue you mentioned in the README file.

In fact, to check whether the issue is caused by the hidden bugs in your pytorch version of ```process.py```, I have adapted your code to the official tensorflow implementation as follows, 
```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import os
from absl import app
from absl import flags
import h5py
import numpy as np
import scipy.io as sio
import tensorflow as tf
import process
import PIL.Image as Image


data_dir = '/media/kaixuan/DATA/Papers/Code/Data/Raw/dnd_2017'
output_dir = './srgb_results'

# Loads image information and bounding boxes.
info = h5py.File(os.path.join(data_dir, 'info.mat'), 'r')['info']
bb = info['boundingboxes']

in_image=tf.placeholder(tf.float32,[None,None,None,4])
sess=tf.Session()

for i in range(1):
  # Loads the noisy image.
  filename = os.path.join(data_dir, 'images_raw', '%04d.mat' % (i + 1))
  img = h5py.File(filename, 'r')

  noisy = np.float32(np.array(img['Inoisy']).T)

  # Loads raw Bayer color pattern.
  bayer_pattern = np.asarray(info[info['camera'][0][i]]['pattern']).tolist()

  boxes = np.array(info[bb[0][i]]).T
  for k in range(20):
    # Crops the image to this bounding box.
    idx = [
        int(boxes[k, 0] - 1),
        int(boxes[k, 2]),
        int(boxes[k, 1] - 1),
        int(boxes[k, 3])
    ]

    noisy_crop = noisy[idx[0]:idx[1], idx[2]:idx[3]].copy()  
    
    # Flips the raw image to ensure RGGB Bayer color pattern.
    if (bayer_pattern == [[1, 2], [2, 3]]):
      pass
    elif (bayer_pattern == [[2, 1], [3, 2]]):
      noisy_crop = np.fliplr(noisy_crop)
    elif (bayer_pattern == [[2, 3], [1, 2]]):
      noisy_crop = np.flipud(noisy_crop)
    else:
      print('Warning: assuming unknown Bayer pattern is RGGB.')

    # Extracts each Bayer image plane.
    height, width = noisy_crop.shape
    channels = []
    for yy in range(2):
      for xx in range(2):
        noisy_crop_c = noisy_crop[yy:height:2, xx:width:2].copy()
        channels.append(noisy_crop_c)
    
    bayer_images = np.stack(channels, axis=-1)[None]

    # Loads shot and read noise factors.
    nlf_h5 = info[info['nlf'][0][i]]
    shot_noise = nlf_h5['a'][0][0]
    read_noise = nlf_h5['b'][0][0]

    # Load the camera's (or image's) ColorMatrix2 
    xyz2cam = tf.constant(np.reshape(np.asarray(info[info['camera'][0][i]]['ColorMatrix2']), (3, 3)), dtype=tf.float32)

    # print(bayer_pattern, xyz2cam)
    # Multiplies with RGB -> XYZ to get RGB -> Camera CCM.
    rgb2xyz = tf.constant([[0.4124564, 0.3575761, 0.1804375],
                            [0.2126729, 0.7151522, 0.0721750],
                            [0.0193339, 0.1191920, 0.9503041]], dtype=tf.float32)
    
    rgb2cam = tf.matmul(xyz2cam, rgb2xyz)
    # Normalizes each row.
    rgb2cam = rgb2cam / tf.reduce_sum(rgb2cam, axis=-1, keepdims=True)

    cam2rgb = tf.matrix_inverse(rgb2cam)
    cam2rgb = tf.expand_dims(cam2rgb, 0)
    # print(cam2rgb, cam2rgb.size())
    # Specify red and blue gains here (for White Balancing)
    asshotneutral = info[info['camera'][0][i]]['AsShotNeutral']
    # print(asshotneutral[1]/asshotneutral[0], asshotneutral[1]/asshotneutral[2])
    red_gain  =  tf.constant(asshotneutral[1]/asshotneutral[0], dtype=tf.float32)
    blue_gain =  tf.constant(asshotneutral[1]/asshotneutral[2], dtype=tf.float32)

    srgb_images = process.process(in_image, red_gain, blue_gain, cam2rgb)

    out_srgb = sess.run(srgb_images, feed_dict={in_image: bayer_images})

    out_srgb = (out_srgb * 255).astype(np.uint8)[0,...]

    # Flip back
    if (bayer_pattern == [[1, 2], [2, 3]]):
      pass
    elif (bayer_pattern == [[2, 1], [3, 2]]):
      out_srgb = np.fliplr(out_srgb)
    elif (bayer_pattern == [[2, 3], [1, 2]]):
      out_srgb = np.flipud(out_srgb)    
    
    Image.fromarray(out_srgb).save(os.path.join(output_dir, '%04d_%02d.png' % (i + 1, k + 1)))
    
```
which utilizes the tensorflow version of ``process.py```.  

However, after this modification, the white balance issue remains. 

We cannot reproduce the rgb results as shown in [here](https://www.timothybrooks.com/tech/unprocessing/darmstadt-supp/)

**Comment by aasharma90** [comment]
Hi @Vandermode , 

Thanks for your feedback. However, I've been quite pressed for time, so I'm sorry I couldn't look into it. I suspect the problem lies within my implementation of `dnd_denoise.py`, more specifically, in the visualization code therein. While coding `process.py` and `unprocess.py`, I took extra care as in I verified them by testing their operations against their corresponding tensorflow counterparts. I could be wrong, but I think the problem does not lie there. 

I'll definitely look into it once I get some time. Meanwhile, if you could figure out it, please do a pull request, and I'll happily merge it. 

Thanks!

**Comment by Vandermode** [comment]
Hi, Thank you for your response. I suspect the issue raised from the inaccurate meta data provided in the DND dataset. In fact,  I have sucessfully adapted your code to postprocess raw-format image captured directly from Sony camera. The output looks very similar to the one processed by ```libraw``` library. 

Thanks!

**Comment by JuZiSYJ** [comment]
Hello
I find the CCM in your code is a little different from the procedure in the  **Rawpy**
Rawpy split the CCM into **cam2xyz** and **xyz2srgb**, while you calculate the **cam2srgb** directly
The difference is
`
    # the code in Rawpy
    raw_rgb = np.array([[0.2,0.25,0.2]])

    cam2xyz1 = np.linalg.inv(color1_norm)
    cam2xyz2 = np.linalg.inv(color2_norm)

    xyz = cam2xyz1 * raw_rgb
    xyz = np.sum(xyz, axis=-1)
    print('first xyz: ', xyz,'\n') #[ 0.19690752  0.23392737  0.16993922] ,default using color_matrix1

    xyz2srgb_norm = xyz2srgb / np.sum(xyz2srgb, axis=-1, keepdims=True)
    srgb = xyz2srgb * xyz
    srgb = np.sum(srgb, axis=-1)
    print('seconed srgb: ', srgb, '\n') # [ 0.19377101  0.2550565   0.14289339]

    ## the code in this repository
    rgb2cam = np.dot(color2, srgb2xyz)
    rgb2cam  = rgb2cam / np.sum(rgb2cam, axis=-1, keepdims=True)

    cam2rgb = np.linalg.inv(rgb2cam)

    srgb2 = raw_rgb * cam2rgb
    srgb2 = np.sum(srgb2, axis=-1)

    print('your way 2 srgb:', srgb2) 
    #[ 0.1511925   0.26764383  0.15978883],using color_matrix1
    #[ 0.17992886  0.26783648  0.17242599], using color_matrix2
`
Hope this will be helpful

**Comment by aasharma90** [comment]
Hi @Vandermode @JuZiSYJ 

Apparently, as reported by @happycaoyue, using the transpose of the bayer pattern fixes this problem. You may have a look at the updated `dnd_denoise.py` file. Thanks!
